pennylane/beta/queuing/measure.py

```diff
@@ -17,16 +17,157 @@
+            if obs is not None:
+                raise ValueError("Cannot set the eigenvalues if an observable is provided.")
+
+            self._eigvals = np.array(eigvals)
+
+        if wires is not None:
+            if obs is not None:
+                raise ValueError("Cannot set the wires if an observable is provided.")
+
+            self._wires = wires
+
+        # TODO: remove the following lines once devices
+        # have been refactored to accept and understand recieving
+        # measurement processes rather than specific observables.
+
+        # The following lines are only applicable for measurement processes
+        # that do no have corresponding observables (e.g., Probability). We use
+        # them to 'trick' the device into thinking it has recieved an observable.
+
+        # Below, we imitate an identity observable, so that the
+        # device undertakes no action upon recieving this observable.
+        self.name = "Identity"
+        self.diagonalizing_gates = lambda: []
+        self.data = []
+
+        # Queue the measurement process
+        self.queue()
+
+    @property
+    def wires(self):
+        r"""The wires the measurement process acts on."""
+        if self.obs is not None:
+            return self.obs.wires
+        return self._wires
+
+    @property
+    def eigvals(self):
+        r"""Eigenvalues associated with the measurement process.
+
+        If the measurement process has an associated observable,
+        the eigenvalues will correspond to this observable. Otherwise,
+        they will be the eigenvalues provided when the measurement
+        process was instantiated.
+
+        Note that the eigenvalues are not guaranteed to be in any
+        particular order.
+
+        **Example:**
+
+        >>> m = MeasurementProcess(Expectation, obs=qml.PauliX(wires=1))
+        >>> U.eigvals
+        >>> array([1, -1])
+
+        Returns:
+            array: eigvals representation
+        """
+        if self.obs is not None:
+            try:
+                return self.obs.eigvals
+            except NotImplementedError:
+                pass
+
+        return self._eigvals
+
+    def expand(self):
+        """Expand the measurement of an observable to a unitary
+        rotation and a measurement in the computational basis.
+
+        Returns:
+            .QuantumTape: a quantum tape containing the operations
+            required to diagonalize the observable
+
+        **Example**
+
+        Consider a measurement process consisting of the expectation
+        value of an Hermitian observable:
+
+        >>> H = np.array([[1, 2], [2, 4]])
+        >>> obs = qml.Hermitian(H, wires=['a'])
+        >>> m = MeasurementProcess(Expectation, obs=obs)
+
+        Expanding this out:
+
+        >>> tape = m.expand()
+
+        We can see that the resulting tape has the qubit unitary applied,
+        and a measurement process with no observable, but the eigenvalues
+        specified:
```

Comment:
Super cool! 

So, will the ``expand()`` function always be called before sending a tape to a device? Or can the device decide? 

I'm thinking about non-qubit devices, for example homodyne measurement, where a computational basis (the Fock basis?) is very inefficient...Would the Gaussian device just ignore the eigenvalues?

Reply:
It depends how we want to proceed!

* We could query the device capabilities, and use that to expand the tape in advance (like we do now with decompositions)

* We could leave the expansion to the device. This might be more elegant? But also more maintenance; changes to expansion would require updates to all the plugins

Reply:
The nice thing about having the `expand` option is it gives multiple options and doesn't force a specific one